### PR TITLE
chore: upgrade prom-client from 11.x to 12.x and update tests.

### PR DESCRIPTION
**Pull request by tarlety**
1. Upgrade prom-client from 11 to 12.
2. Update tests for prom-client 12.0.0: ```metrics are collected on scrape of metrics endpoint, not on an interval.```
3. Update tests for prom-client 12.0.0: ```Cannot add the default metrics twice to the same registry```
4. Refactor integration tests based on different server/configuration.

### CODE REVIEW

File: src/metrics-middleware.js [hunk 1]
```diff
@@ -65,6 +56,17 @@ module.exports = (appVersion, projectName, framework = 'express') => {
     };
 };
 
+function PromehteusRegisterAppVersion(appVersion, metricName) {
```
Comment (thosil) [inline src/metrics-middleware.js]:
```suggestion
function PrometheusRegisterAppVersion(appVersion, metricName) {
```

File: src/metrics-middleware.js [hunk 2]
```diff
@@ -28,16 +28,7 @@ module.exports = (appVersion, projectName, framework = 'express') => {
...
-            const version = new Prometheus.Gauge({
-                name: metricNames.app_version,
-                help: 'The service version by package.json',
-                labelNames: ['version', 'major', 'minor', 'patch']
-            });
-
-            const versionSegments = appVersion.split('.').map(Number);
-            version.labels(appVersion, versionSegments[0], versionSegments[1], versionSegments[2]).set(1);
-        }
+        PromehteusRegisterAppVersion(appVersion, metricNames.app_version);
```
Comment (thosil) [inline src/metrics-middleware.js]:
```suggestion
        PrometheusRegisterAppVersion(appVersion, metricNames.app_version);
```

File: src/metrics-middleware.js (hunk 1 above)
Comment (tarlety) [reply src/metrics-middleware.js]: fix in 1bab09d

File: src/metrics-middleware.js (hunk 2 above)
Comment (tarlety) [reply src/metrics-middleware.js]: fix in 1bab09d

**Review by thosil — COMMENTED** [review]
Looks good ~~, only a small typos~~

File: src/express-middleware.js [hunk 3]
```diff
@@ -5,7 +5,7 @@ const utils = require('./utils');
 
 class ExpressMiddleware {
     constructor(setupOptions) {
-        this.setupOptions = setupOptions || {};
+        this.setupOptions = setupOptions;
```
Comment (kobik) [inline src/express-middleware.js:8]: why removing the default value?

File: src/metrics-middleware.js [hunk 4]
```diff
@@ -65,6 +56,17 @@ module.exports = (appVersion, projectName, framework = 'express') => {
...
 };
 
+function PrometheusRegisterAppVersion(appVersion, metricName) {
+    const version = new Prometheus.Gauge({
+        name: metricName,
+        help: 'The service version by package.json',
+        labelNames: ['version', 'major', 'minor', 'patch']
+    });
+
+    const versionSegments = appVersion.split('.').map(Number);
```
Comment (kobik) [inline src/metrics-middleware.js]: we can destructure for more readability `const [ major, minor, patch ] = appVersion.split('.').map(Number);`

File: src/metrics-middleware.js [hunk 5]
```diff
@@ -65,6 +56,17 @@ module.exports = (appVersion, projectName, framework = 'express') => {
     };
 };
 
+function PrometheusRegisterAppVersion(appVersion, metricName) {
```
Comment (kobik) [inline src/metrics-middleware.js:59]: we previously had this condition `if (!Prometheus.register.getSingleMetric(metricNames.app_version))`. why removing it?

**Review by kobik — COMMENTED** [review]
thanks @tarlety , please see my comments

File: src/express-middleware.js (hunk 3 above)
Comment (tarlety) [reply src/express-middleware.js:8]: Here are my thoughts:
1. Only ```frameworkMiddleware()``` (in metrics-middleware.js) passes setupOptions to ```ExpressMiddleware``` (in express-middleware.js), and setupOptions is always being constructed well.
2. No code uses the default value.
3. ```KoaMiddleware``` (in koa-middleware.js) does not have this default value either.
4. To increase the test coverage.

File: src/metrics-middleware.js (hunk 4 above)
Comment (tarlety) [reply src/metrics-middleware.js]: You are right. It makes the code more readable.

The refactoring is done: 89ce2e3

File: src/metrics-middleware.js (hunk 5 above)
Comment (tarlety) [reply src/metrics-middleware.js:59]: With prom-client 12.x, ```metrics-middleware``` cannot be called twice without prometheus.register.clear().

Constructing middleware twice will cause Prometheus.collectDefaultMetrics() to throw exception  in prom-client/lib/defaultMetrics.js:
```js=48
  if (last) {
    throw new Error(
      'Cannot add the default metrics twice to the same registry'
    );
  }
```

Some test cases are modified in commit b5cac62: ```fix: npm test fail due to 'Cannot add the default metrics twice to the same registry'```
1. It always calling prometheus.register.clear() before next test.
2. New test case: ```Cannot add the default metrics twice to the same registry```. It expects a exception for second middleware function.

**Comment by kobik** [comment]
We also need to update `engines` in package.json to node >= 10 and that would be a good time to add a CHANGELOG.md file before releasing this version,

File: .travis.yml [hunk 6]
```diff
@@ -1,6 +1,5 @@
 language: node_js
 node_js:
-  - "8"
   - "10"
   - "12"
```
Comment (kobik) [inline .travis.yml:4]: can you please add 14?

**Comment by kobik** [comment]
As we have here a breaking change,I also guess it would be a good time to add a CHANGELOG.md file before releasing this version.

It would nice to add `engines` in package.json with node >= 10.

File: .travis.yml (hunk 6 above)
Comment (tarlety) [reply .travis.yml:4]: Ok, done in 10547b5.

**Review by kobik — APPROVED** [review]

**Pull request merged**